encode.go

```diff
@@ -38,82 +39,134 @@ func (e Encoder) Encode(v interface{}) (err error) {
 	if err = e.encodeMapValueMaybe(); err != nil {
 		return
 	}
-	if v == nil {
+
+	// This type switch optimizes encoding of common value types, it prevents
+	// the use of reflection to identify the type of the value, which saves a
+	// dynamic memory allocation.
+	switch x := v.(type) {
+	case nil:
 		return e.Emitter.EmitNil()
-	}
-	return e.encode(reflect.ValueOf(v))
-}
 
-// EncodeBool uses e to encode the boolean value v.
-func (e Encoder) EncodeBool(v bool) (err error) {
```

Comment:
Breaking change? If so, should we leave them in for compatibility and add a deprecation note?